modules/navier_stokes/src/fvkernels/PINSFVMomentumPressureFlux.C

```diff
@@ -0,0 +1,64 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PINSFVMomentumPressureFlux.h"
+#include "PINSFVSuperficialVelocityVariable.h"
+#include "NS.h"
+
+registerMooseObject("NavierStokesApp", PINSFVMomentumPressureFlux);
+
+InputParameters
+PINSFVMomentumPressureFlux::validParams()
+{
+  auto params = FVFluxKernel::validParams();
+  params.addClassDescription("Viscous diffusion term, div(mu grad(u)), as a flux kernel "
+                             "using the divergence theoreom, in the porous media "
+                             "incompressible Navier-Stokes momentum equation. This kernel "
+                             "is also executed on boundaries.");
+  params.addRequiredCoupledVar("porosity", "Porosity auxiliary variable");
+  params.addRequiredCoupledVar("p", "Pressure variable");
+  MooseEnum momentum_component("x=0 y=1 z=2");
+  params.addParam<MooseEnum>("momentum_component",
+                             momentum_component,
+                             "The component of the momentum equation that this kernel applies to.");
+  params.set<unsigned short>("ghost_layers") = 2;
+  params.set<bool>("force_boundary_execution") = true;
+  return params;
+}
+
+PINSFVMomentumPressureFlux::PINSFVMomentumPressureFlux(const InputParameters & params)
+  : FVFluxKernel(params),
+    _eps(coupledValue("porosity")),
+    _p_elem(adCoupledValue(NS::pressure)),
+    _p_neighbor(adCoupledNeighborValue(NS::pressure)),
+    _index(getParam<MooseEnum>("momentum_component"))
+{
+#ifndef MOOSE_GLOBAL_AD_INDEXING
+  mooseError("PINSFV is not supported by local AD indexing. In order to use PINSFV, please run "
+             "the configure script in the root MOOSE directory with the configure option "
+             "'--with-ad-indexing-type=global'");
+#endif
+  if (!dynamic_cast<PINSFVSuperficialVelocityVariable *>(&_var))
+    mooseError("PINSFVMomentumPressureFlux may only be used with a superficial velocity, "
+               "of variable type PINSFVSuperficialVelocityVariable.");
+}
+
+ADReal
+PINSFVMomentumPressureFlux::computeQpResidual()
+{
+  ADReal p_interface;
+
+  Moose::FV::interpolate(Moose::FV::InterpMethod::Average,
+                         p_interface,
+                         _p_elem[_qp],
+                         _p_neighbor[_qp],
+                         *_face_info,
+                         true);
+  return _eps[_qp] * p_interface * _normal(_index);
```

Comment:
At one point in time you interpolated epsilon and pressure together. Now you are not. Why is that? This way is not conservative. The neighbor porosity is not entering into the calculation at all...I'm a little confused as to why the change.

Reply:
The term is eps * grad_P or eps * div (**I** P). So I take out (assumed constant) eps from the volume integral before applying the divergence theorem.

To have eps * P evaluated at the faces, we would need div(eps P **I**), which we could make appear with an IBP but then there's a porosity gradient which isnt great either.

Reply:
The concern I guess is that depending on which element is chosen when the flux kernel is evaluated on the face, the contribution to the residual will be different. That could explain the difficulties in parallel.

I ll change that